experiment-trait-data-visualizer/app.R

```diff
@@ -1,163 +1,272 @@
+    render_date <- input [[ paste0('map_date_', season_name) ]]
+    
+    traits <- full_cache_data[[ season_name ]][[ 'trait_data' ]][[ selected_variable ]][[ 'traits' ]]
+    
+    if (selected_cultivar != 'None')
+      traits <- subset(traits, cultivar_name == selected_cultivar)
+    
+    units <- full_cache_data[[ season_name ]][[ 'trait_data' ]][[ selected_variable ]][[ 'units' ]]
+    if (units != '')
+      units <- paste0('(', units, ')')
+    legend_title <- paste0(selected_variable, ' ', units)
+    
+    render_site_map(traits, render_date, legend_title)
   })
+}
+
+# render outputs for a given season
+render_season_output <- function(season_name, input, output, full_cache_data) {
+  
+  render_variable_menu(season_name, output, full_cache_data)
+  
+  render_cultivar_menu(season_name, input, output, full_cache_data)
+  
+  render_trait_plot(season_name, input, output, full_cache_data)
+  
+  render_mgmt_timeline(season_name, input, output, full_cache_data)
   
-  # render plot for selected variable
-  output$traitPlot <- renderPlot({
+  render_plot_hover(season_name, input, output, full_cache_data)
+  
+  render_timeline_hover(season_name, input, output, full_cache_data)
 
-    # get access to 'fullTraitData' from cache
-    data.cache(cache.name=cacheName(), loadTraitData, startDate=seasonStartDate(), endDate=seasonEndDate())
-    
-    # only render plot of a variable is selected
-    if (!is.null(input$selectedVariable)) {
-      # get observations for selected variable
-      variableIdData <- betydb_query(table='variables', id=input$selectedVariable)
-      variableTraitData <- subset(fullTraitData, variable_id==as.numeric(variableIdData$id))
-      
-      # get specie data for title
-      specieId <- unique(as.numeric(variableTraitData$specie_id))
-      specieData <- betydb_query(table='species', id=specieId)
-      title <- paste0('Mean ', gsub('_', ' ', variableIdData$name), ' for ', specieData$scientificname)
+  render_map(season_name, input, output, full_cache_data)
+}
 
-      # generate timeseries of boxplots from mean value
-      ggplot(variableTraitData, aes(as.Date(date), mean)) + 
-      geom_boxplot(aes(group=cut_width(as.Date(date), 1))) +
-      labs(title=title,
-        x="Observation Dates", y=variableIdData$units) +
-        theme(text = element_text(size=20), axis.text.x = element_text(angle=45, hjust=1)) +
-        expand_limits(y=0) + xlim(seasonStartDate(), seasonEndDate())
-    }
-  })
+server <- function(input, output) {
   
-  # generate timeline visualization for managements data
-  output$timeline <- renderTimevis({
-    mgmtData <- getManagementsData(startDate=seasonStartDate(), endDate=seasonEndDate())
-    timelineData <- data.frame(
-      id=1:nrow(mgmtData),
-      content=paste0(mgmtData$mgmttype),
-      start=as.Date(mgmtData$date)
-    )
-    timevis(timelineData)
+  # load 'full_cache_data' object from cache file
+  load('cache.RData')
```

Comment:
A better solution would be to check if the file exists and if not, to run the script to generate it.